File: apps/desktop/src/lib/redact.ts
/**
 * Secret redaction utilities.
 * Applies a regex pack over session text before it leaves the machine.
 */
import type { NormalizedSession, SessionMessage } from "./types";

// ─── regex pack ──────────────────────────────────────────────────────────────

const REDACT_PLACEHOLDER = "[REDACTED]";

interface RedactRule {
  label: string;
  pattern: RegExp;
}

const RULES: RedactRule[] = [
  // AWS access key IDs
  { label: "aws-key-id", pattern: /\bAKIA[0-9A-Z]{16}\b/g },
  // AWS secret access keys (40 char base64-ish after assignment)
  {
    label: "aws-secret",
    pattern: /(?:aws_secret_access_key|AWS_SECRET_ACCESS_KEY)\s*=\s*["']?[A-Za-z0-9/+]{40}["']?/gi,
  },
  // Anthropic / OpenAI-style sk- tokens
  { label: "sk-token", pattern: /\bsk-[A-Za-z0-9_-]{20,}/g },
  // Slack tokens xox*
  { label: "xox-token", pattern: /\bxox[bopas]-[0-9A-Za-z\-]{10,}/g },
  // Bearer tokens in Authorization headers
  {
    label: "bearer-token",
    pattern: /(?:Authorization|authorization)\s*[:=]\s*["']?Bearer\s+[A-Za-z0-9\-._~+/]+=*["']?/g,
  },
  // Generic API key / secret assignments  (KEY=val, SECRET=val, TOKEN=val, PASSWORD=val)
  {
    label: "env-secret",
    pattern: /\b(?:API_KEY|API_SECRET|SECRET_KEY|TOKEN|PASSWORD|PASSWD|AUTH_TOKEN)\s*=\s*["']?[^\s"']{8,}["']?/g,
  },
  // PEM private key blocks
  {
    label: "pem-block",
    pattern: /-----BEGIN (?:RSA |EC |OPENSSH )?PRIVATE KEY-----[\s\S]*?-----END (?:RSA |EC |OPENSSH )?PRIVATE KEY-----/g,
  },
  // JWTs (three base64url segments separated by dots)
  {
    label: "jwt",
    pattern: /\beyJ[A-Za-z0-9_-]+\.eyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+/g,
  },
];

// ─── core text redactor ───────────────────────────────────────────────────────

export function redactText(text: string): { text: string; count: number } {
  let count = 0;
  let result = text;
  for (const rule of RULES) {
    const replaced = result.replace(rule.pattern, () => {
      count++;
      return REDACT_PLACEHOLDER;
    });
    result = replaced;
  }
  return { text: result, count };
}

// ─── session redactor ────────────────────────────────────────────────────────

function redactMessage(msg: SessionMessage): { msg: SessionMessage; count: number } {
  let totalCount = 0;

  const { text: newText, count: c1 } = redactText(msg.text);
  totalCount += c1;

  let newThinking = msg.thinking;
  if (msg.thinking) {
    const { text: t2, count: c2 } = redactText(msg.thinking);
    newThinking = t2;
    totalCount += c2;
  }

  return {
    msg: { ...msg, text: newText, thinking: newThinking },
    count: totalCount,
  };
}

export function redactSession(
  s: NormalizedSession
): { session: NormalizedSession; count: number } {
  let totalCount = 0;

  const redactedMessages = s.messages.map((m) => {
    const { msg, count } = redactMessage(m);
    totalCount += count;
    return msg;
  });

  const { text: newPreview, count: pc } = redactText(s.preview);
  totalCount += pc;

  const { text: newTitle, count: tc } = redactText(s.title);
  totalCount += tc;

  return {
    session: {
      ...s,
      title: newTitle,
      preview: newPreview,
      messages: redactedMessages,
    },
    count: totalCount,
  };
}

File: apps/api/contexthub/ingest/redact.py
"""Secret redaction for session text.

Applies a regex pack that catches the most common accidental secret leaks:
  - AWS access key IDs (AKIA…)
  - Generic API tokens (sk-…, xox.-…, ghp_…, etc.)
  - Bearer tokens in Authorization headers
  - PEM private key blocks
  - .env-style KEY=value assignments
  - JWTs (three base64url segments separated by dots)
  - Email addresses (optional, off by default)

Returns the redacted text and a count of replacements made.
"""

from __future__ import annotations

import re
from typing import Optional

# Each pattern: (compiled_regex, replacement_string)
_PATTERNS: list[tuple[re.Pattern[str], str]] = [
    # AWS access key ID  (20-char uppercase alphanumeric starting with AKIA)
    (
        re.compile(r"\bAKIA[0-9A-Z]{16}\b"),
        "[REDACTED:AWS_KEY_ID]",
    ),
    # AWS secret access key (40-char base64-ish, often follows the key id)
    (
        re.compile(r"(?i)aws[_\-]?secret[_\-]?access[_\-]?key\s*[=:]\s*\S+"),
        "[REDACTED:AWS_SECRET]",
    ),
    # Generic sk- tokens (OpenAI, Anthropic, Stripe, etc.)
    (
        re.compile(r"\bsk-[A-Za-z0-9_\-]{20,}\b"),
        "[REDACTED:API_TOKEN]",
    ),
    # Slack / Xoxp tokens
    (
        re.compile(r"\bxox[baprs]-[A-Za-z0-9\-]{10,}\b"),
        "[REDACTED:SLACK_TOKEN]",
    ),
    # GitHub personal access tokens (classic ghp_ or fine-grained github_pat_)
    (
        re.compile(r"\b(?:ghp|github_pat)_[A-Za-z0-9_]{20,}\b"),
        "[REDACTED:GITHUB_TOKEN]",
    ),
    # Bearer tokens in Authorization headers
    (
        re.compile(r"(?i)(Authorization\s*:\s*Bearer\s+)\S+"),
        r"\1[REDACTED:BEARER_TOKEN]",
    ),
    # PEM private key blocks
    (
        re.compile(
            r"-----BEGIN (?:RSA |EC |OPENSSH )?PRIVATE KEY-----.*?-----END (?:RSA |EC |OPENSSH )?PRIVATE KEY-----",
            re.DOTALL,
        ),
        "[REDACTED:PEM_PRIVATE_KEY]",
    ),
    # .env-style assignments: SECRET_KEY=value, API_KEY="value", etc.
    (
        re.compile(
            r"(?i)(?:password|secret|api[_\-]?key|auth[_\-]?token|access[_\-]?token|private[_\-]?key)"
            r"\s*[=:]\s*['\"]?\S+['\"]?",
        ),
        "[REDACTED:ENV_SECRET]",
    ),
    # JWTs — three base64url segments separated by dots
    (
        re.compile(
            r"\beyJ[A-Za-z0-9_\-]+\.[A-Za-z0-9_\-]+\.[A-Za-z0-9_\-]+\b"
        ),
        "[REDACTED:JWT]",
    ),
]

# Optional email redaction pattern (not in the default pack)
_EMAIL_PATTERN = re.compile(r"\b[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}\b")


def redact_text(text: str, redact_emails: bool = False) -> tuple[str, int]:
    """Apply the full redaction regex pack to *text*.

    Args:
        text: The input string, possibly containing secrets.
        redact_emails: Whether to also redact email addresses (default off).

    Returns:
        A (redacted_text, count) tuple where count is the number of
        individual substitutions performed.
    """
    count = 0
    result = text
    for pattern, replacement in _PATTERNS:
        new_result, n = pattern.subn(replacement, result)
        count += n
        result = new_result

    if redact_emails:
        new_result, n = _EMAIL_PATTERN.subn("[REDACTED:EMAIL]", result)
        count += n
        result = new_result

    return result, count
